### PR TITLE
seeed/beaglev: first commit

### PR DESCRIPTION
**Pull request by rminnich**
includes clock.rs for the starchip jh7100 and some basic structure.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

**Comment by rminnich** [comment]
rustfmt does not fix stuff like gofmt does. Cargo fix does nothing so the useless parens are there. This was taken from C code for starfive.

**Review by sbanikintel — DISMISSED** [review]
Care to change the copyright year ?

>> Copyright (C) 2018 Jonathan Neuschäfer

File: src/mainboard/seeed/beaglev/DIT [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+dd if=target/riscv64imac-unknown-none-elf/release/image.bin  of=image.bin bs=80K count=1
+
+ls -l image.bin
+sudo strace -o shit -f -s 1024 -v \
+     /home/rminnich/bin/jh7100-recover \
+     -D /dev/ttyUSB0 \
+     -r /home/rminnich/riscv/vic_second_boot.bin \
+     -b  image.bin \
+     -d /home/rminnich/riscv/ddrinit-2133-210302.bin.out
```
Comment (alistair23) [inline src/mainboard/seeed/beaglev/DIT:20]: This isn't going to work for anyone else

File: src/mainboard/seeed/beaglev/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
+#![feature(llvm_asm)]
```
Comment (alistair23) [inline src/mainboard/seeed/beaglev/src/main.rs:1]: Do you actually need this?

File: src/mainboard/seeed/beaglev/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+pub mod uart;
+
+global_asm!(include_str!(
+    "../../../../../src/soc/starfive/jh7100/src/bootblock.S"
+));
+global_asm!(include_str!(
+    "../../../../../src/soc/starfive/jh7100/src/init.S"
+));
+
+// TODO: For some reason, on hardware, a1 is not the address of the dtb, so we hard-code the device
```
Comment (alistair23) [inline src/mainboard/seeed/beaglev/src/main.rs:29]: Do we expect the hardware to pass a DT to us?

File: src/mainboard/seeed/beaglev/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+// tree here. TODO: The kernel ebreaks when given this device tree.
```
Comment (alistair23) [inline src/mainboard/seeed/beaglev/src/main.rs:30]: The DT should be updated from the kernel source then

File: src/soc/starfive/jh7100/src/clock.rs [hunk 5]
```diff
@@ -4505,19 +4505,19 @@ impl<'a> Clock<'a> {
...
+        // TODO: make base a parameter.
+        _SWITCH_CLOCK_clk_cpundbus_root_SOURCE_clk_pll0_out_();
+        _SWITCH_CLOCK_clk_dla_root_SOURCE_clk_pll1_out_();
+        _SWITCH_CLOCK_clk_dsp_root_SOURCE_clk_pll2_out_();
+        _SWITCH_CLOCK_clk_perh0_root_SOURCE_clk_pll0_out_();
+
+        // not enabled in original.
+        // slow down nne bus can fix nne50 & vp6 ram scan issue,
+        // as well as vin_subsys reg scan issue.
+        //	_SWITCH_CLOCK_clk_nne_bus_SOURCE_clk_cpu_axi_;
```
Comment (alistair23) [inline src/soc/starfive/jh7100/src/clock.rs]: Why change the indentation?

File: azure-pipelines.yml [hunk 6]
```diff
@@ -28,8 +28,8 @@ jobs:
       make --keep-going test
     displayName: 'Run rust tests'
   - script: |
-      make --keep-going clippy
-    displayName: 'Run clippy linter'
+      echo let us wait until cargo fix or rustfix actually works make --keep-going clippy
+    displayName: 'SKIP clippy linter'
```
Comment (rjoleary) [inline azure-pipelines.yml:32]: Can you create a github issue for this to remind us?

File: src/mainboard/seeed/beaglev/DIT [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+sudo date
+
+make
+
+echo "power on, hold down reset, once jh7100 starts"
+
+dd if=target/riscv64imac-unknown-none-elf/release/image.bin  of=image.bin bs=80K count=1
+
+ls -l image.bin
+sudo strace -o shit -f -s 1024 -v \
```
Comment (rjoleary) [inline src/mainboard/seeed/beaglev/DIT:15]: Why are you running it with strace?

File: src/mainboard/seeed/beaglev/DIT [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+set -e
+
+echo empty sudo so we do not need another password prompt
+sudo date
+
+make
+
+echo "power on, hold down reset, once jh7100 starts"
+
+dd if=target/riscv64imac-unknown-none-elf/release/image.bin  of=image.bin bs=80K count=1
```
Comment (rjoleary) [inline src/mainboard/seeed/beaglev/DIT:12]:
```suggestion
dd if=target/riscv64imac-unknown-none-elf/release/image.bin of=image.bin bs=80K count=1
```

File: src/mainboard/seeed/beaglev/DIT [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (rjoleary) [inline src/mainboard/seeed/beaglev/DIT:1]: This should probably be in the Makefile.

File: src/mainboard/seeed/beaglev/Makefile [hunk 10]
```diff
@@ -0,0 +1,8 @@
+TARGET     = riscv64imac-unknown-none-elf
+QEMU       ?= qemu-system-riscv64
+QEMU_FLAGS += -m 1g -machine sifive_u,start-in-flash=true -nographic -device loader,addr=0x20000000,file=${IMAGE} -bios none -smp 4
```
Comment (rjoleary) [inline src/mainboard/seeed/beaglev/Makefile:3]: The machine type is still sifive_u, is this fine?

File: src/mainboard/seeed/beaglev/fixed-dtfs.dts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+/ {
+    #address-cells = <1>;
+    #size-cells = <1>;
+
+    flash-info {
+        compatible = "ore-flashinfo";
+        board-name = "HiFive Unleashed";
+        category = "SiFive";
+        board-url = "https://www.sifive.com/boards/hifive-unleashed";
```
Comment (rjoleary) [inline src/mainboard/seeed/beaglev/fixed-dtfs.dts:11]: Change these three lines.

File: src/mainboard/seeed/beaglev/DIT (hunk 1 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/DIT:20]: good point :-)

File: src/soc/starfive/jh7100/src/clock.rs (hunk 5 above)
Comment (rminnich) [reply src/soc/starfive/jh7100/src/clock.rs]: rustfmt did that one.

File: src/mainboard/seeed/beaglev/src/main.rs (hunk 2 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/src/main.rs:1]: yeah. there's an include. Well, we used to need it, do we not need it?

File: src/mainboard/seeed/beaglev/src/main.rs (hunk 3 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/src/main.rs:29]: the bootblock passes it and on the beagle there is a vic_secondboot that *may* pass it. Still not sure.

File: src/mainboard/seeed/beaglev/src/main.rs (hunk 4 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/src/main.rs:30]: that comment may be wrong ....

appreciate the review very much. You can see this is quite a WIP, a.k.a. POS

File: azure-pipelines.yml (hunk 6 above)
Comment (rminnich) [reply azure-pipelines.yml:32]: will do.

File: src/mainboard/seeed/beaglev/DIT (hunk 7 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/DIT:15]: it's a terrible program. It hangs, and ... well, lots of things. I'll probably be removing this file at some point.

File: src/mainboard/seeed/beaglev/DIT (hunk 9 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/DIT:1]: yep.

File: src/mainboard/seeed/beaglev/Makefile (hunk 10 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/Makefile:3]: it's a weird machine. ALl their docs reference that core.

File: src/mainboard/seeed/beaglev/Makefile [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+include ../../../../Makefile.inc
+include ../../../../Makefile.qemu.inc
+
+openocd:
+	$(shell which openocd) -f openocd.cfg -c "flash write_image erase unlock $(IMAGE) 0x20000000; shutdown"
```
[the remark this answers is not in the document]
Comment (rminnich) [reply src/mainboard/seeed/beaglev/Makefile:8]: I *think* maybe soon?

File: src/mainboard/seeed/beaglev/fixed-dtfs.dts (hunk 11 above)
Comment (rminnich) [reply src/mainboard/seeed/beaglev/fixed-dtfs.dts:11]: will do.

**Comment by rminnich** [comment]
Thanks everyone for reviewing this preliminary piece of coprolite.

File: src/mainboard/seeed/beaglev/src/uart.rs [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    serial_out(
+        REG_FCR,
+        (FCR_FIFO | FCR_MODE1 | /*FCR_FIFO_1*/FCR_FIFO_8 | FCR_RCVRCLR | FCR_XMITCLR),
+    );
+
+    serial_out(REG_IER, 0); //dis the ser interrupt
+}
+
+fn x4(data: u8) {
+    match data {
```
Comment (rjoleary) [inline src/mainboard/seeed/beaglev/src/uart.rs:172]: Could you do something like `putc("0123456789abcdef"[data])`?

**Comment by rminnich** [comment]
This board is dead, no need to bring it in I think.

**Comment by fotonick** [comment]
You mean that your board let out the magical smoke or that the manufacturer has discontinued the line?

**Review by orangecms — DISMISSED** [review]
let's settle on this :boat:

File: src/mainboard/seeed/beaglev/Makefile (hunk 10 above)
Comment (orangecms) [inline src/mainboard/seeed/beaglev/Makefile:3]: just put an empty task at the very top; every board is built in CI - the lists for skipping are only for formatting and crate testing. So something here like this shall do:

```Makefile
default:
    echo "Skipping Beagle V"
```

**Review by orangecms — DISMISSED** [review]
please be nice, Azure

**Review by orangecms — DISMISSED** [review]

**Review by orangecms — APPROVED** [review]

**Pull request merged**